crates/vizia_core/src/views/textbox.rs

```diff
@@ -527,52 +470,62 @@ where
     }
 
     fn new_core(cx: &mut Context, lens: L, kind: TextboxKind) -> Handle<Self> {
+        // TODO can this be simplified now that text doesn't live in TextboxData?
         let result = Self { lens: lens.clone(), kind }.build(cx, move |cx| {
             Binding::new(cx, lens.clone(), |cx, text| {
-                let text =
-                    text.get_fallible(cx).map(|x| x.to_string()).unwrap_or_else(|| "".to_owned());
+                let text_str = text.view(cx.data().unwrap(), |text| {
+                    text.map(|x| x.to_string()).unwrap_or_else(|| "".to_owned())
+                });
                 if let Some(text_data) = cx.data::<TextboxData>() {
                     if !text_data.edit {
                         let td = TextboxData {
-                            text: text.clone(),
-                            selection: text_data.selection,
+                            //text: text.clone(),
+                            //selection: text_data.selection,
                             edit: text_data.edit,
-                            sel_x: text_data.sel_x,
-                            re_sel_x: text_data.re_sel_x,
+                            //sel_x: text_data.sel_x,
+                            //re_sel_x: text_data.re_sel_x,
                             transform: text_data.transform,
-                            line_height: text_data.line_height,
+                            //line_height: text_data.line_height,
                             on_edit: text_data.on_edit.clone(),
                             content_entity: text_data.content_entity,
                             kind: text_data.kind,
                             on_submit: text_data.on_submit.clone(),
                         };
+                        cx.cosmic_context.with_buffer(text_data.content_entity, |buf| {
+                            buf.set_text(&text_str, Attrs::new());
+                        });
                         let parent = cx.current().parent(&cx.tree).unwrap();
                         cx.with_current(parent, |cx| td.build(cx));
                         // push an event into the queue to force an update because the textbox data
                         // may have already been observed this update cycle
                         cx.emit_to(cx.current(), ());
                     }
                 } else {
-                    let mut td = TextboxData::new(text.clone());
+                    let mut td = TextboxData::new();
                     td.set_caret(&mut EventContext::new(cx));
                     let parent = cx.current().parent(&cx.tree).unwrap();
                     cx.with_current(parent, |cx| td.build(cx));
                     cx.emit_to(cx.current(), ());
                 }
             });
+            let text = lens.view(cx.data().unwrap(), |text| {
+                text.map(|x| x.to_string()).unwrap_or_else(|| "".to_owned())
+            });
             TextboxContainer {}
```

Comment:
Do we still need a container as well as the content?

Reply:
I think we could _technically_ get rid of container, it would just be more computation to apply the margins between the textbox and the container. I'm fine having the layout system compute that for us.